### PR TITLE
bumblebee-Jenny Xu

### PR DESCRIPTION
**Pull request by jennylearncoding**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={sender === 'Vladimir' ? 'chat-entry local' : 'chat-entry remote'}>
+      <h1>{sender}</h1>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:12]: Nice use of the supplied `TimeStamp`. We pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,31 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id : PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLikeToggle: PropTypes.func.isRequired
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:28]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  entries: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.string.isRequired,
+      liked: PropTypes.bool.isRequired,
+    })
+  ).isRequired,
+  onLikeToggle: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:33]: Similar to the props for `ChatEntry`, here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import './ChatLog.css';
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry';
+
+const ChatLog = ({ entries, onLikeToggle}) => {
+  return (
+    <div className="chat-log">
+      {entries.map((entry) => (
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:8]: Nice use of `map` to convert from the message data into `ChatEntry` components. We could perform this mapping storing the result into a variable we use in the JSX result, we could make a helper function that we call as part of the `return`, or we can perform the map expression as part of the `return` JSX as you did here. I often like having relatively simple maps directly in the JSX since it helps me see the overall structure of the component, though it can make debugging a little more tricky. But any of those approaches will work fine.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        <ChatEntry 
+          key={entry.id}
+          id={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:11]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+          sender={entry.sender}
+          body={entry.body}
+          timeStamp={entry.timeStamp}
+          liked={entry.liked || false}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:15]: 🔎 I'm curious what the `false` fallback value is doing here. `liked` is included as part of the message data, meaning this should be defined for every message.

File: src/components/ChatEntry.jsx [hunk 7]
```diff
@@ -1,20 +1,31 @@
...
+      <h1>{sender}</h1>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
+        </p>
+        <button className="like" onClick={() => onLikeToggle(id)}>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:14]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 8]
```diff
@@ -1,20 +1,31 @@
...
+          {liked ? '❤️' : '🤍'}
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:15]: 👍 Nice use of a ternary to pick the emoji to show for the liked status. Consider moving even something small like this to a helper function external to the component, so that the behavior of the emoji-picking logic can be tested easily outside of React.

File: src/App.jsx [hunk 9]
```diff
@@ -1,17 +1,32 @@
...
+import { useState } from 'react';
 
 const App = () => {
+  const [chat, setChat] = useState(messages);
+
+  const togglelike = (id) => {
+    const updatedMessages = chat.map((msg) =>
+      msg.id === id ? { ...msg, liked: !msg.liked } : msg
+    );
+    setChat(updatedMessages);
```
Comment (anselrognlie) [inline src/App.jsx:13]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setChat(chat => chat.map(msg => {
      if (msg.id === id) {
        return {...msg, liked: !msg.liked};
      } else {
        return msg;
      };
    }));
```
We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 10]
```diff
@@ -1,17 +1,32 @@
...
+  };
+
+  const total = chat.filter((msg) => msg.liked).length;
```
Comment (anselrognlie) [inline src/App.jsx:16]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Great idea to filter down the data to the liked messages and use the list length as the count. This is a really understandable alternative to the more canonical `reduce` approach. Compared to `reduce`, it does make a list potentially as long as the message list, but since we know a similar copying gets done anyway during a render (building the `ChatEntry` components) on the whole it's not any more costly.

File: src/components/ChatEntry.jsx [hunk 11]
```diff
@@ -1,20 +1,31 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLikeToggle }) => {
   return (
-    <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={sender === 'Vladimir' ? 'chat-entry local' : 'chat-entry remote'}>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: Nice logic to decide whether to treat a message as local or remote. How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatEntry` itself have enough information as it is to "know" which messages are local and which are remote?

File: src/components/ChatEntry.jsx [hunk 12]
```diff
@@ -1,20 +1,31 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLikeToggle }) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params), but I really prefer the glanceability.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
